---
title: "Sea-urchin-Upwelling-HW2"
output: html_document
---


Title: Examining Differential Gene Expression in Strongylocentrotus purpuratus 
(purple sea urchin) under upwelling vs. non-upwelling conditions. 

Authors: 
Victoria French
Madison Pacaro 
Jamie Poirier 


# Introduction

For this assignment, we used data presented in ["Examining the Role of DNA
Methylation in Transcriptomic Plasticity of Early Stage Sea Urchins: Developmental and Maternal Effects in a Kelp Forest Herbivore"](https://www.frontiersin.org/articles/10.3389/fmars.2020.00205/full) by M. Strader et al. (2020) in Frontiers in Marine Science.

In Strader et al.â€™s work, they explored the relationship between the dynamics of gene expression and the effects of epigenetic markers. It considers the extent of environmental effects on plasticity both within and between generations of sea urchins. They worked with Stronglyocentrotus purpuratus and simulated both upwelling and non-upwelling conditions, and examined connections in changes regarding DNA methylation, transcription, and spicule length of their offspring during embryonic development.   

![*Stronglyocentrotus purpuratus*](C:/Users/jamie/OneDrive/Documents/bi586/assignment2/GO_enrichment/sea_urchin_upwelling/S.purpuratus.jpg)
Figure 1: Stronglyocentrotus purpuratus: photo via. Wikipedia (Kirt L. Onthank)

In their analysis, they used reduced representation bisulfite sequencing (RRBS) sequencing, pre-processing and analysis, which is a high-throughput technique used for analysis of methylation within a genome. Following this, they used RNA sequencing as well as differential gene expression analysis and weighted gene co-expression network analysis (WGCNA), which is used for the correlation and clustering of genes.

They ultimately found that forms of genetic environmentally induced plasticity are similar within groups of genes, but on an individual level, these changes were largely unique. It also found that there were significant differences in methylation of CpG sites and gene expression based on maternal conditioning, and environmental conditions during embryonic development had a significant effect on spicule length. 


# Version Control


R version 4.0.3 was used for all analyses. 
DESeq2 package version 1.30.1
affycoretools package version 1.62.0
arrayQualityMetrics package version 3.46.0
genefilter package version 1.72.1
Biobase package version 2.50.0
dplyr package version 1.0.5
RColorBrewer package version 1.1-2
gplots package version 3.1.1
vegan package version 2.5-7
ggplot 2 package version 3.3.3
ggrepel package version 0.9.1
tidyverse package version 1.3.0
pheatmap package version 1.0.12


# Data Analysis & Methods

First, we loaded all the required packages for our analyses.
```{r, warning=FALSE, message=FALSE}

#DESeq 
library(DESeq2) 
library(affycoretools)
library(arrayQualityMetrics)
library(genefilter)
library(Biobase)
library(dplyr)
library(RColorBrewer)
library(gplots)
library(vegan)
library(ggplot2)
library(ggrepel)
library(tidyverse)
library(pheatmap)

#library(VennDiagram)

#GO


```

We then set our working directory for the project.
```{r, results='hide'}
setwd("C:/Users/Maddy/Documents/BI586/seaurchinupwelling")

```

Next, we read in our raw count data, and determined the length of the data. 
```{r}
countData <- read.table("geneCounts_02122019.txt")
head(countData)
length(countData[,1])
```

The length of our data is 30284. This is the number of genes we have represented in our data set.

We then remaned our data by our two treatments: NN = non-upwelling and UU = upwelling conditions.
```{r}
names(countData)=c("NN", "NN", "NN", "UU", "UU", "UU")
head(countData)
```


# Outlier Analysis

We then conducted array quality metrics to detect and remove outliers in our data. We first specified a new directory for outlier data and our treatments.
```{r}
setwd("C:/Users/Maddy/Documents/BI586/seaurchinupwelling/outlier")
v=setwd("C:/Users/Maddy/Documents/BI586/seaurchinupwelling/outlier")
treat=c("NN", "NN", "NN", "UU", "UU", "UU")
```

Here, we are creating colData, which is a data frame associating sample with treatment.
```{r}
g=data.frame(treat)
g
colData=g
```

We now are calling on DESeq2 to create a model for our data, which is storing the results of analysis of differential expression and asking how our design varies by treatment. We then normalized our data.
```{r}

dds=DESeqDataSetFromMatrix(countData=countData,
                           colData = g,
                           design = ~treat)
vsd.ge=assay(vst(dds))
rl=vst(dds)
```

Using arrayQualityMetrics, we then identified any outliers in our data. An outlier in UU was detected under "Distances between arrays." We will be keeping this outlier in our data because it was only offensive in distance between arrays, but not in boxplots or MA plots. Based on the barplot of our total counts data below, we will see even distribution between our treatments, which is another reason we are keeping this outlier.

#show html file here #link? #https://scc-ondemand1.bu.edu/rnode/scc-wi3/26578/files/assignment2/seaurchinupwelling/index.html

```{r}
e=ExpressionSet(assay(rl), AnnotatedDataFrame(as.data.frame(colData(rl))))
arrayQualityMetrics(e,outdir=v,intgroup=c("treat"),force=T)
```

We then changed directories to our original directory.
```{r}
setwd("C:/Users/Maddy/Documents/BI586/seaurchinupwelling")
```

We then analyzed our count data by looking at the total number of counts. From this output, we can see that our raw counts range from approximately 6 million to 8 million reads across treatments. We then plotted a bar plot of the counts which shows that our raw counts are generally uniformly distributed for each treatment, which is a good signal for normalization. 
```{r}
totalCounts=colSums(countData)

totalCounts
barplot(totalCounts, col=c("coral", "coral", "coral", "red", "red", "red"), ylab="raw counts")

min(totalCounts) 
max(totalCounts)  
```

We then created the DESeq object and saved the results of the model.
```{r}
dds<-DESeq(dds)

head(dds) 
res<- results(dds)
res
```

Then, to look at the dispersion of our data, we plotted a visual representation of the DESeq object. This plot follows the general shape of a hockey stick, which shows that the dispersion fits well to the curve.
```{r}
plotDispEsts(dds, main="Dispersion plot")
```


# Treatment Comparisons

Here, we perform analyses to examine differentially expressed genes among our two treatment groups: upwelling vs. non-upwelling. In the first analysis, we specified non-upwelling (NN) as our control in the levels option and our results will therefore be in terms of upwelling compared to non-upwelling. 

We looked at signficantly differentially expressed genes in upwelling conditions at verying levels of significance. At the highest confidence level we have 715 differentially expressed genes in upwelling vs. non-upwelling conditions. 

```{r}
colData$UU<-factor(colData$treat, levels=c("UU","NN"))
resUU <- results(dds, contrast=c("treat","UU","NN"))
#how many FDR < 10%?
table(resUU$padj<0.001) 
# 0.1=3039 - not valuable
# 0.05=2289
# 0.01=1344
# 0.001 =715

```

We summarized resUU. This allows us to see that there are genes removed in pairwise analysis due to low counts (23%) and we also see 5.4% of genes were downregulated in UU relative to NN. In addition, 7.2% of genes were upregulated in UU relative to NN.
```{r}
summary(resUU)
```

Here we are using another method to look at differentially expressed genes, as we did above. We see the number of significantly differentially expressed genes excluding low count genes.
```{r}
nrow(resUU[resUU$padj<0.05 & !is.na(resUU$padj),])   
```

We plotted a visual representation of differentially expressed genes and this shows........ *add explanation
```{r}
plotMA(resUU, main="NN vs UU") 
```

We summarized our results.
```{r}
results <- as.data.frame(resUU)
head(results)
```

Then, we examined upregulated and downregulated genes separately. Upregulated genes are genes with a positive logfold change and downregulated genes are genes with a negative logfold change.
```{r}
nrow(resUU[resUU$padj<0.1 & resUU$log2FoldChange > 0 & !is.na(resUU$padj),])
nrow(resUU[resUU$padj<0.1 & resUU$log2FoldChange < 0 & !is.na(resUU$padj),])
#UP in UU is 1735
#DOWN in UU is 1304
```

We created a table to summarize our results from above. 
```{r}
write.table(resUU, file="UU_DEG.txt", quote=F, sep="\t") 

cd <- read.table("UU_DEG.txt")
cd

```

We created a GO table with ranked p-values with directionality for each gene. We then saved the table as a CSV for future GO analysis.
```{r}
cd
go_input_UU = cd %>%
  tibble::rownames_to_column(var = "iso") %>%
  mutate(mutated_p = -log(pvalue)) %>%
  mutate(mutated_p_updown = ifelse(log2FoldChange < 0, mutated_p*-1, mutated_p*1)) %>%
  na.omit() %>%
  select(iso, mutated_p_updown)

head(go_input_UU)
colnames(go_input_UU) <- c("gene", "pval")
head(go_input_UU)
write.csv(go_input_UU, file="UU_GO.csv", quote=F, row.names=FALSE)
#in outlier folder
```

We bound the p-value from res comparisons and the p-adjusted values. 

```{r}
valUU=cbind(resUU$pvalue, resUU$padj)
head(valUU)
#this is pvalues and padjusted values
colnames(valUU)=c("pval.UU", "padj.UU")
length(valUU[,1]) #this is the number of genes we are looking at = 30284 (same as above)
table(complete.cases(valUU))
#False = NAs
```

Making log fold change and p-value data table 

Doing r log transformation, this is important for making heat maps. it is a normalization method and is also how we are going to make PCAs 

```{r}
rlog=rlogTransformation(dds, blind=TRUE) 
rld=assay(rlog)
head(rld)
#this shows us for each isogroup, the r log normalized values for each of our samples
colnames(rld)=paste(colData$treat)
head(rld)
length(rld[,1]) #length should be still same, indicating retention of all data

#binding rld data and pvalues in order to rank p-values for top 100 heatmap later on 
rldpvals=cbind(rld,valUU)
head(rldpvals)
dim(rldpvals) #looking at dimensions [1] 30284     8 , there are more columns here because we added columns of pvalues
table(complete.cases(rldpvals))
#FALSE  TRUE 
#11745  18539 , sanity check: we still have the same number of NAs (false) here

write.csv(rldpvals, "RLDandPVALS.csv", quote=F)


```

Creating the Sample Distance heatmap 

```{r}
rldpvals <- read.csv(file="RLDandPVALS.csv", row.names=1)
head(rldpvals)
rld=rldpvals[,1:6] 
#making rld which is just columns 1-6 and will cut off pvalues to just leave us with r log normalized isogroups
head(rld)

sampleDists <- dist(t(rld))
sampleDistMatrix <- as.matrix( sampleDists )
treat=c( "upwelling", "upwelling", "upwelling", "non-upwelling", "non-upwelling", "non-upwelling")
colnames(sampleDistMatrix)=paste(treat)
rownames(sampleDistMatrix)=paste(treat)

#creating a heat map from the sampledistMatrix 
heat.colors = colorRampPalette(rev(c("navy","beige")),bias=0.3)(100)
pheatmap(sampleDistMatrix,color = heat.colors,cex=0.9,border_color=NA,cluster_rows=T,cluster_cols=T)

#heatmap shows clustering - how similar samples are 
#these are a good way to visualize overall expression between our samples
#this heatmap looks good, it is showing the clustering of our treatments, which we expect 
```

Principle Component Analysis 
```{r}
rld_t=t(rld) #transposing data frame 

#Removing columns with zero variance (all log fold change 0 meaning not differentially expressed)
which(apply(rld_t, 2, var)==0)
rld_t <- rld_t[ , which(apply(rld_t, 2, var) != 0)]

pca <- prcomp(rld_t,center = TRUE, scale. = TRUE) #up to 6 principle components because n=6 samples 
head(pca)

#we are interested in PC1 and PC2 bc these are the two principle components that explain the most variance

#defining amount of variance described by PC1 and PC2 
li <- pca$sdev^2 / sum(pca$sdev^2)
pc1v <- round(li[1] * 100, 1)
pc2v <- round(li[2] * 100, 1)
pca_s <- as.data.frame(pca$x)
head(pca_s)
pca_s <- pca_s[,c(1,2)] #pulling out first two PCs
pca_s$Samples = row.names(pca_s)
pca_s$treat=colData$treat
head(pca_s)

#Plotting pca with ggplot
cbPalette <- c("darkgoldenrod2",  "darkolivegreen3", "dodgerblue3")
ggplot(pca_s, aes(PC1, PC2, color = treat, pch = treat)) +
  geom_point(size=3) +
  #  geom_text_repel(aes(label=Samples)) +
  scale_colour_manual(values=cbPalette)+
  theme_bw() +
  # geom_density2d(alpha=.5)+
  geom_polygon(alpha=.2)+
  xlab(paste0("PC1: ",pc1v,"% variance")) +
  ylab(paste0("PC2: ",pc2v,"% variance")) 
head(pca)
```

Statistical Analysis ############## Double check this significance; it was significant when first ran and now isn't?  #########

```{r}
#ANOVA analysis testing if distances between samples on PCA are significantly different 
#treatment (at pvalue 0.003) has a specific impact on gene differential expression 
adonis(pca$x ~ treat, data = pca_s, method='eu', na.rm = TRUE)
```

Heatmap for top 100 most significant differently expressed genes

```{r}
rldpvals <- read.csv(file="RLDandPVALS.csv", row.names=1)
head(rldpvals)
rld_site= rldpvals[,1:6] #isolating vsd data columns 
head(rld_site)
gg=read.table("goAnnot_spu.tab",sep="\t", row.names=1)
head(gg)

nrow(rldpvals[rldpvals$padj.UU<0.01& !is.na(rldpvals$padj.UU),])
#1344; isogroups extremely differently expressed (p-adjusted value <0.01) 

#Ranking p-adjusted values to find top 100 most differentially expressed genes 
topnum= 100 # number of DEGS in final heatmap 
head(rldpvals)
top100=head(rldpvals[order(rldpvals$padj.UU), ],topnum)
head(top100)
length(top100[,1])
summary(top100)


head(top100)
p.val=0.1 # FDR cutoff
conds=top100[top100$padj.UU<=p.val & !is.na(top100$padj.UU),] #all top 100 where the p- adjusted value is less than or equal to .1 (significant)
length(conds[,1])

exp=conds[,1:6] #removing associated p-values from log fold change data
means=apply(exp,1,mean) # means of rows
explc=exp-means # subtracting them
colnames(explc)=c("NN1", "NN2", "NN3", "UU1", "UU2", "UU3")
head(explc)

ccol=colorRampPalette(rev(c("red","chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)
col0=colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)

pheatmap(explc,cluster_cols=T,scale="row",color=col0, show_rownames = F)
```

Adding Gene annotation to top 100 heatmap 

```{r}
ann = data.frame(cond = c('UU', 'UU', 'UU', 'NN', 'NN', 'NN'))
rownames(ann) <- names(explc)

# Set colors
Var1        <- c("darkgoldenrod2",  "darkolivegreen3", "black")
names(Var1) <- c("UU", "NN")
anno_colors <- list(cond = Var1)

pheatmap(as.matrix(explc),annotation_col=ann,annotation_colors=anno_colors,cex=.85,color=col0,border_color=NA,clustering_distance_rows="correlation",clustering_distance_cols="correlation", show_rownames=T)
```

#Figures

```{r}
#Figure 1
cbPalette <- c("darkgoldenrod2",  "darkolivegreen3", "dodgerblue3")
ggplot(pca_s, aes(PC1, PC2, color = treat, pch = treat)) +
  geom_point(size=3) +
  #  geom_text_repel(aes(label=Samples)) +
  scale_colour_manual(values=cbPalette)+
  theme_bw() +
  # geom_density2d(alpha=.5)+
  geom_polygon(alpha=.2)+
  ggtitle("Gene Expression")+
  theme(plot.title = element_text(hjust = 0.5))+
  xlab(paste0("PC1: ",pc1v,"% variance")) +
  ylab(paste0("PC2: ",pc2v,"% variance")) 
```
Figure 1: Principal component analysis of gene expression. % variance explained by principal component specified on respective axes labels. Treatment replicates differentiated by shape and color according to legend. NN=Non-upwelling conditions. UU=upwelling conditions 

```{r}
#Figure 2
pheatmap(as.matrix(explc),annotation_col=ann,annotation_colors=anno_colors,cex=.85,color=col0,border_color=NA,clustering_distance_rows="correlation",clustering_distance_cols="correlation", show_rownames=T)
```
Figure 2: Differentially expressed genes by treatment with associated gene annotation. Scale of expression is log2fold change relative to the gene mean. Condition key specifies treatment. UU= upwelling conditions NN= non-upwelling conditions NA=Undefined 

```{r}
#Figure 3 #GO 
```



#Conclusion 



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```




